rfcs/80-curve-trait.md

```diff
@@ -0,0 +1,793 @@
+where
+    Self: Sized,
+{ //... }
+```
+As you can see, `reparametrize` is like `map`, but the function is applied in parameter space instead of in
+output space. This is somewhat counterintuitive, because it means that many of the functions that we might want
+to use it with actually need to be inverted. For example, here is how you would use `reparametrize` to change
+the `domain` of a curve from `[0.0, 1.0]` to `[0.0, 2.0]` by linearly stretching it:
+```rust
+let my_curve = function_curve(interval(0.0, 1.0).unwrap(), |x| x + 1.0);
+let domain = my_curve.domain();
+let scaled_curve = my_curve.reparametrize(interval(0.0, 2.0).unwrap(), |t| t / 2.0);
+```
+(A convenience method `reparametrize_linear` exists for this specific kind of thing.)
+
+However, `reparametrize` can do much more than this alone. For example, here we use `reparametrize` to
+create a new curve that is a segment of the original one:
+```rust
+let my_curve = function_curve(interval(0.0, 1.0).unwrap(), |x| x * 2.0);
+// The segment of `my_curve` from `0.5` to `1.0`, shifted back to `0.0:
+let curve_segment = my_curve.reparametrize(interval(0.0, 0.5).unwrap(), |t| 0.5 + t);
+```
+
+And here, we use it to reverse our curve:
+```rust
+let my_curve = function_curve(interval(0.0, 2.0).unwrap(), |x| x * x);
+let domain = my_curve.domain();
+let reversed_curve = my_curve.reparametrize(domain, |t| domain.end() - t);
+```
+
+And here, we reparametrize by an easing curve:
+```rust
+let my_curve = function_curve(interval(0.0, 1.0).unwrap(), |x| x + 5.0);
+let easing_curve = function_curve(interval(0.0, 1.0).unwrap(), |x| x * x * x);
+let eased_curve = my_curve.reparametrize(easing_curve.domain(), |t| easing_curve.sample(t));
+```
+(The latter also has a convenience method, `reparametrize_by_curve`, which handles the domain automatically.)
+
+---
+
+Next, we have `graph`, the last of the general functional methods of the API:
+```rust
+/// Create a new [`Curve`] which is the graph of this one; that is, its output includes the
+/// parameter itself in the samples. For example, if this curve outputs `x` at time `t`, then
+/// the produced curve will produce `(t, x)` at time `t`.
+fn graph(self) -> impl Curve<(f32, T)>
+where
+    Self: Sized,
+{ //... }
+```
+This is a subtle method whose main applications involve allowing more complex things to be expressed. For example,
+here we modify a curve by making its output value attenuate with time:
+```rust
+// A curve with a value of `3.0` over the `Interval` from `0.0` to `5.0`:
+let my_curve = const_curve(interval(0.0, 5.0).unwrap(), 3.0);
+// The same curve but with exponential falloff in time:
+let new_curve = my_curve.graph().map(|(t, x)| x * (-t).exp2());
+```
+
+Most general operations that one could think up for curves can be achieved by clever combinations of `map` and
+`reparametrize`, perhaps with a little `graph` sprinkled in.
+
+### Sampling, resampling, and interpolation
```

Comment:
> As Curve's can be created from functions, which do not have control points, the Curve API does not give access to information about the control points of a curve. This section does not introduce additional traits that would give access to such information, right?

Not for the trait itself, but the intent is for the control data to be largely transparent to the user (and serialized/deserialized and so on). 

> In that case I would suggest a trait `Invertible` (or `Bijective`) for `Curve<f32>` that adds a method `invert(&self) -> Curve<f32>` such that for all `x` in `curve.domain()` the invariant `x = curve.invert().sample(curve.sample(x))` holds. (Or perhaps a `try_invert`, which might be more meaningful for the use case below). Note that a curve that implements this must be strictly increasing or decreasing.

I'm a bit skeptical of basing this on a trait, since that requires a type-level guarantee of invertibility or something, which seems rarer; my personal view is that it's more likely that we want to implement some inversion techniques at the level of library functions, since it's easier to do that than to try to track some kind of invertibility through a bunch of transformations. 

> Use case: You have a `timestamp: Vec<f32>` and `keyframe: Vec<f32>` of equal length. You want a curve that interpolates over these keyframes, respecting the timestamps. If timestamp is strictly increasing, then `SampleCurve::new(timestamp).invert().chain(SampleCurve(keyframes))` would be that curve.

> This might result in a simpler implementation than with `EvenCore` and `UnevenCore`.

Eh, I think both of these deserve to exist. Even if, in principle, you can accomplish one using the other, it sounds like you are going to end up storing the same data anyway, and this gets even hairier when you consider the fact that interpolation is not always scale-invariant; this shows up, for instance, in the [glTF VariableCurve rewrite, with cubic interpolation](https://github.com/bevyengine/bevy/pull/13105). Regardless, I think keyframe interpolation is basic and useful enough to deserve its own data structures.